cmd/update-rules/main.go

```diff
@@ -0,0 +1,162 @@
+	outExample := "-o /tmp/rules.yaml"
+
+	branchUsage := "Branch of k/k to create rules for, for example: " + branchExample
+	ruleUsage := "Path to rule file to update rules for, for example: " + ruleExample
+	goVerUsage := "Golang version to pin for this branch, if not specified default go version from input rule file is used, for example: " + goVerExample
+	outUsage := "Output the updated rules file at given path instead of updating the input rules file, for example: " + outExample
+	var examples = `
+  Examples:
+  # Update rules and replace input rule file for release-1.22 branch with default go version from input rule file pinned
+  update-rules -branch release-1.22 -rule /go/src/k8s.io/kubernetes/staging/publishing/rules.yaml
+
+  # Update rules and export to /tmp/updated_rules.yaml for release-1.22 branch with go version version 1.16.3
+  update-rules -branch release-1.22 -go 1.16.3 -o /ymp/updated_rules.yaml -rule /go/src/k8s.io/kubernetes/staging/publishing/rules.yaml`
+
+	flag.Usage = func() {
+		fmt.Fprintf(os.Stdout, "\n  Usage:  update-rules %s %s [%s | %s]", branchExample, ruleExample, "-go", "-o")
+		fmt.Fprintf(os.Stdout, "\n  %s\n\n", examples)
+		flag.PrintDefaults()
+	}
+
+	branch := flag.String("branch", "", branchUsage)
+	rulesFile := flag.String("rule", "", ruleUsage)
+	goVersion := flag.String("go", "", goVerUsage)
+	out := flag.String("o", "", outUsage)
+	// TODO: add dryrun option
+
+	flag.Parse()
+
+	if *branch == "" {
+		glog.Errorf("branch flag requires a non-empty value, for example: %s", branchExample)
+		os.Exit(2)
+	}
+	// TODO: add branch validation
+
+	if *rulesFile == "" {
+		glog.Errorf("rule flag requires a non-empty value, for example: %s", ruleExample)
+		os.Exit(2)
+	}
+
+	var outFile string
+	if *out != "" {
+		outFile = *out
+	} else {
+		outFile = *rulesFile
+	}
+
+	rules, err := config.LoadRules(*rulesFile)
+	if err != nil {
+		glog.Fatalf("can not load rules file %q: %v", *rulesFile, err)
+	}
+
+	if err := config.Validate(rules); err != nil {
+		glog.Fatalf("invalid rule file %q: %v", *rulesFile, err)
+	}
+
+	// use the golang version specified via input or default in input rule file
+	if *goVersion == "" {
+		*goVersion = *rules.DefaultGoVersion
+	}
+
+	for j, r := range rules.Rules {
+		var mainBranchRuleFound bool
+		var newBranchRule config.BranchRule
+		// find the mainBranch rule
+		for _, br := range r.Branches {
+			if br.Name == mainBranch {
+				newBranchRule = config.BranchRule(br)
+				mainBranchRuleFound = true
+				break
+			}
+		}
+		// if mainBranch rule not found for target repo, error
+		if !mainBranchRuleFound {
+			glog.Fatalf("%s branch rule not found for repo %s", mainBranch, r.DestinationRepository)
+		}
+
+		// copy the mainBranch rule and update the branch name and go version
+		newBranchRule.Name = *branch
+		newBranchRule.Source.Branch = *branch
+		// pin the go version per branch rule update
+		if newBranchRule.GoVersion == "" {
+			newBranchRule.GoVersion = *goVersion
+		}
```

Comment:
I think we can unconditionally update the go version here, without needed to check if a go version was specified for the master branch

Reply:
changed to only update the go version if specified via CLI flag, else leave untouched / omitempty